02 Charging For Your App/app.rb

```diff
@@ -0,0 +1,207 @@
+require 'shopify_api'
+require 'sinatra'
+require 'httparty'
+require 'dotenv'
+
+class GiftBasket < Sinatra::Base
+  attr_reader :key,:secret,:tokens
+
+  def initialize
+    Dotenv.load
+    @key = ENV['API_KEY']
+    @secret = ENV['API_SECRET']
+    @app_url = "jamie.ngrok.io"
+    @tokens = {}
+    super
+  end
+
+  get '/giftbasket/install' do
+    shop = request.params['shop']
+    scopes = "read_orders,read_products,write_products"
+
+    # construct the installation URL and redirect the merchant
+    install_url = "http://#{shop}/admin/oauth/authorize?client_id=#{@key}"\
+                "&scope=#{scopes}&redirect_uri=https://#{@app_url}/giftbasket/auth"
+
+    # redirect to the install_url
+    redirect install_url
+  end
+
+  get '/giftbasket/auth' do
+    # extract shop data from request parameters
+    shop = request.params['shop']
+    code = request.params['code']
+    hmac = request.params['hmac']
+
+    # perform hmac validation to determine if the request is coming from Shopify
+    validate_hmac(hmac,request)
+
+    # if no access token for this particular shop exist,
+    # POST the OAuth request and receive the token in the response
+    get_shop_access_token(shop,key,secret,code)
+
+    # now that the session is activated, we can create a recurring application charge
+    create_recurring_application_charge
+
+    # we want to redirect to the bulk edit URL if there is a token and an activated session and an activated RecurringApplicationCharge
+    redirect bulk_edit_url
+  end
+
+  get '/activatecharge' do
+    # store the charge_id from the request
+    charge_id  = request.params['charge_id']
+    recurring_application_charge = ShopifyAPI::RecurringApplicationCharge.find(charge_id)
+    recurring_application_charge.status == "accepted" ? recurring_application_charge.activate : return # how to redirect back to confirmation URL so they loop until accept payment
```

Comment:
you bet.

![02-39-3s2k3-2s3nf](https://cloud.githubusercontent.com/assets/10549158/18209488/eb062f70-7101-11e6-8a4e-9312b8cb0f9b.jpg)